mlir/test/mlir-pdll/CodeGen/MLIR/expr.pdll

```diff
@@ -142,3 +142,99 @@ Pattern RangeExpr {
 // CHECK: %[[TYPE:.*]] = type : i32
 // CHECK: operation({{.*}}) -> (%[[TYPE]] : !pdl.type)
 Pattern TypeExpr => erase op<> -> (type<"i32">);
+
+// -----
+
+//===----------------------------------------------------------------------===//
+// Parse attributes and rewrite
+//===----------------------------------------------------------------------===//
+
+// Rewriter helpers declarations.
+Rewrite createDictionaryAttr() -> Attr;
+Rewrite addEntryToDictionaryAttr(dict: Attr, attrName: Attr, attr : Attr) -> Attr;
+Rewrite createArrayAttr() -> Attr;
+Rewrite addElemToArrayAttr(arrayAttr: Attr, newElement: Attr) -> Attr;
```

Comment:
Outside this example, there are no pointers to the need to provide the PDLL declarations, nor to provide documentation and/or implementation of what they do and instructions to link and register the C++ implementations of them when one uses this feature.

Agreed to a kind of "level 0" support where you describe the approach in the description of this PR, and add comments to state that it requires a C++ implementation to be linked and registered in passes that process PDLL with "feature X".